### PR TITLE
ci(NODE-6505): Setup CI

### PR DESCRIPTION
**Pull request by aditi-khare-mongoDB**
**Descriptions**
This testing framework is being added in to allow for ease of encryption testing in PRs to come.

Add CI configuration to ensure that FLE can be end-to-end tested in the mongoose repo.
Add scripting so users can test locally and understand how the CI configuration works.

**Summary of Changes**
- Add scripting to setup CI for mongoose integration tests locally and through GHA
- Add script to setup encryption cluster locally for local user testing
- Add basic integration testing to check the CI has been setup correctly

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -22,6 +22,7 @@
     "bson": "^6.7.0",
     "kareem": "2.6.3",
     "mongodb": "~6.10.0",
+    "mongodb-client-encryption": "^6.1.0",
```
Comment (aditi-khare-mongoDB) [inline package.json:25]:
```suggestion
```

File: .github/workflows/encryption-tests.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: Encryption Tests 
+
+on:
+  push
+  #workflow_dispatch: {}
```
Comment (baileympearson) [inline .github/workflows/encryption-tests.yml]: Intentionally commented out?

File: test/encryption/encryption.test.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
+'use strict';
+
+const assert = require('assert');
+const mdb = require('mongodb');
+
+describe('environmental variables', () => {
+  it('MONGODB_TEST_URI is set', async function() {
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]:
```suggestion
  it('MONGOOSE_TEST_URI is set', async function() {
```

File: test/encryption/encryption.test.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+    const client = new mdb.MongoClient(
+      process.env.MONGOOSE_TEST_URI,
+      {
+        autoEncryption: {
+          keyVaultNamespace: 'keyvault.datakeys',
+          kmsProviders: { local: { key: Buffer.from(
+            'Mng0NCt4ZHVUYUJCa1kxNkVyNUR1QURhZ2h2UzR2d2RrZzh0cFBwM3R6NmdWMDFBMUN3YkQ5aXRRMkhGRGdQV09wOGVNYUMxT2k3NjZKelhaQmRCZGJkTXVyZG9uSjFk',
+            'base64'
+          )
+          } },
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: The formatting here looks like it got a bit messed up.  Is lint running against this file?

File: test/encryption/encryption.test.js [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+});
+
+describe('basic integration', () => {
+  it('supports mongodb csfle auto-encryption integration', async() => {
+    // 1. Create a MongoClient configured with auto encryption (referred to as `client_encrypted`)
+    const client = new mdb.MongoClient(
+      process.env.MONGOOSE_TEST_URI,
+      {
+        autoEncryption: {
+          keyVaultNamespace: 'keyvault.datakeys',
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: Without a schemaMap (or an encryptedFieldsMap), we won't actually encrypt anything.  We need to provide a schema map or encrypted fields map that specifies that `encrypted` should be encrypted - otherwise we're just inserting a document into the cluster.

File: .github/workflows/encryption-tests.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+        run: npm install mongodb-client-encryption
+      - name: Set up cluster
+        id: setup-cluster
+        uses: mongodb-labs/drivers-evergreen-tools@master
+        with:
+          version: ${{ matrix.mongodb }}
+          topology: sharded_cluster 
+          auth: auth
+      - name: Run Tests
+        run: npx mocha --exit ./test/encryption/*.test.js
```
Comment (baileympearson) [inline .github/workflows/encryption-tests.yml]: Usually, we'd add an npm script that runs the tests instead of invoking mocha directly

File: .github/workflows/encryption-tests.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      security-events: write
+      id-token: write
+      contents: write
+    runs-on: ${{ matrix.os }}
+    strategy:
+      fail-fast: false
+      matrix:
+        node: [20]
+        os: [ubuntu-latest]
+        mongodb: [8.0.0]
```
Comment (baileympearson) [inline .github/workflows/encryption-tests.yml]: A testing matrix isn't actually necessary when you're only testing on one variant of each matrix axis (we're always testing node 20, ubuntu latest, server 8.0) - so we can just hard code these values into the script wherever they're needed.

File: .github/scripts/run-kms-servers.sh [hunk 8]
```diff
@@ -0,0 +1,11 @@
+#!/bin/bash
+set -o errexit  # Exit the script with error if any of the commands fail
```
Comment (baileympearson) [inline .github/scripts/run-kms-servers.sh]: Are you using this anywhere?

**Review by baileympearson — CHANGES_REQUESTED** [review]
This is a great start! 

Along with the comments, we also need the ability to set up a cluster and run tests locally.  Can we add a script that uses drivers-evergreen-tools to set up a cluster for local development as well as using the github action?

File: .github/scripts/run-kms-servers.sh (hunk 8 above)
Comment (aditi-khare-mongoDB) [reply .github/scripts/run-kms-servers.sh]: Removed the file!

File: .github/workflows/encryption-tests.yml (hunk 7 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: Removed it!

File: .github/workflows/encryption-tests.yml (hunk 6 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: Done! Added a script instead and call that script

File: test/encryption/encryption.test.js (hunk 5 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: Changed it to use a schema map, and checked that binary is inserted in!

File: test/encryption/encryption.test.js (hunk 4 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: The lint is running against this file (if I put a .only in this file, it throws an error)! I changed up the formatting though

File: .github/workflows/encryption-tests.yml (hunk 2 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: Just so any push triggers the action, I've added this branch name into the push config for clarity  - will remove it when we're ready to merge.

File: .github/workflows/encryption-tests.yml [hunk 9]
```diff
@@ -0,0 +1,45 @@
+name: Encryption Tests 
+
+on:
+  push: 
+    branches: ['master', 'NODE-6505/ci-setup']
```
Comment (baileympearson) [inline .github/workflows/encryption-tests.yml]: From the AC on the ticket:

> Run GHA on PR open to main and commit to main, maybe remove the PR trigger after FLE work is done.

We're missing a trigger when PRs are opened against main (if we make that change, you can remove your branch from this list too)

File: mongocryptd.pid
Comment (baileympearson) [inline mongocryptd.pid:1]: This should be deleted.  Should we perhaps added `.pid` files to the .gitignore? (also if we're using crypt_shared correctly, we shouldn't be spawning mongocryptds 🤔 )

File: package.json (hunk 1 above)
Comment (baileympearson) [inline package.json]: This probably shouldn't be a production dependency of mongoose 🙂

File: test/encryption/encryption.test.js [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+          kmsProviders: { local: { key: LOCAL_KEY } },
+          schemaMap: {
+            'db.coll': {
+              bsonType: 'object',
+              encryptMetadata: {
+                keyId: [new mdb.UUID(dataKey)]
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: I think the return type of createDataKey is `UUID`
```suggestion
                keyId: [dataKey]
```

File: test/encryption/encryption.test.js [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    await dummyClient.close();
+  });
+
+  it('supports mongodb csfle auto-encryption integration', async() => {
+    await encryptedClient.connect();
+    await encryptedClient.db('db').collection('coll').insertOne({ a: 1 });
+
+    // a dummyClient not configured with autoEncryption, returns a encrypted binary type, meaning that encryption succeeded
+    const encryptedCursor = await dummyClient.db('db').collection('coll').find();
+    const encryptedResult = await encryptedCursor.next();
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]:
```suggestion
    const { insertedId } = await encryptedClient.db('db').collection('coll').insertOne({ a: 1 });

    // a dummyClient not configured with autoEncryption, returns a encrypted binary type, meaning that encryption succeeded
    const encryptedCursor = await dummyClient.db('db').collection('coll').findOne({ _id: insertedId });
```

(optional) Usually we'd just query for the document we inserted specifically, instead of creating a cursor and iterating it.  This also applies below, to the encryptedClient.find()

**Review by baileympearson — CHANGES_REQUESTED** [review]
I think this was missed from my initial review:

> Along with the comments, we also need the ability to set up a cluster and run tests locally. Can we add a script that uses drivers-evergreen-tools to set up a cluster for local development as well as using the github action?

@aditi-khare-mongoDB

File: mongocryptd.pid
Comment (aditi-khare-mongoDB) [reply mongocryptd.pid:1]: I've deleted the extraneous `.pid` files. They're a result of locally running the tests without a crypt_shared library path correctly set - with the new local encrypted test runner, this should be fixed.

File: package.json (hunk 1 above)
Comment (aditi-khare-mongoDB) [reply package.json]: Removed it!

File: test/encryption/encryption.test.js (hunk 11 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: Made the change. Thanks for this suggestion, puts it into context where I can avoid extra `.next()` calls.

File: .github/workflows/encryption-tests.yml (hunk 9 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: Yep, added it in for `master` on mongoose.

File: test/encryption/encryption.test.js (hunk 10 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: Seems like I was mistaken during our original conversation - this code change works fine - i changed it in the `properties` section as well.

File: test/encryption/encryption.test.js [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+
+  afterEach(async function() {
+    await keyVaultClient.close();
+    await encryptedClient.close();
+    await dummyClient.close();
+  });
+
+  it('supports mongodb csfle auto-encryption integration', async() => {
+    await encryptedClient.connect();
+    await encryptedClient.db('db').collection('coll').insertOne({ a: 1 });
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]:
```suggestion
```

File: test/encryption/encryption.test.js [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    // a dummyClient not configured with autoEncryption, returns a encrypted binary type, meaning that encryption succeeded
+    const encryptedResult = await dummyClient.db('db').collection('coll').findOne({ _id: insertedId });
+
+    assert.ok(encryptedResult);
+    assert.ok(encryptedResult.a);
+    assert.ok(isBsonType(encryptedResult.a, 'Binary'));
+    assert.ok(encryptedResult.a.sub_type === 6);
+
+    // when the encryptedClient runs a find, the original unencrypted value is returned
+    const unencryptedCursor = await encryptedClient.db('db').collection('coll').find();
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: We should use findOne here as well

File: test/encryption/encryption.test.js [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+          },
+          extraOptions: {
+            cryptdSharedLibRequired: true,
+            cryptSharedLibPath: process.env.CRYPT_SHARED_LIB_PATH
+          }
+        }
+      }
+    );
+
+    dummyClient = new mdb.MongoClient(process.env.MONGOOSE_TEST_URI);
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: Can we use a better name than `dummyClient` - maybe `unencryptedClient`?

File: test/encryption/encryption.test.js [hunk 15]
```diff
@@ -0,0 +1,97 @@
...
+  });
+
+  afterEach(async function() {
+    await keyVaultClient.close();
+    await encryptedClient.close();
+    await dummyClient.close();
+  });
+
+  it('supports mongodb csfle auto-encryption integration', async() => {
```
Comment (baileympearson) [inline test/encryption/encryption.test.js]: Can we either use a more descriptive test name or leave a comment here explaining what the test is intended for?

Specifically, the intention behind this test is to just confirm that we've set up the mongoose FLE testing environment.

File: scripts/start-encrypted-cluster.sh [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+
+if [ ! -d "drivers-evergreen-tools/" ]; then
+ git clone --depth=1 "https://github.com/mongodb-labs/drivers-evergreen-tools.git"
+fi
+
+export DRIVERS_TOOLS=$CWD/encrypted-cluster/drivers-evergreen-tools
+export MONGODB_VERSION=8.0
+export AUTH=true
+export MONGODB_BINARIES=$DRIVERS_TOOLS/mongodb/bin
+export NODE_DRIVER=~/dev/node-mongodb-native
```
Comment (baileympearson) [inline scripts/start-encrypted-cluster.sh]:
```suggestion
```

File: scripts/run-encryption-tests-local.sh [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+
+export CWD=$(pwd);
+
+if [ -d "encrypted-cluster" ]; then
+  cd encrypted-cluster
+else
+  source $CWD/scripts/start-encrypted-cluster.sh
+fi
+
+# IMPORTANT: extracts mongodb-uri, and starts the cluster of servers, store the uri for GitHub output
```
Comment (baileympearson) [inline scripts/run-encryption-tests-local.sh]: Is this comment relevant anymore?  We're not using this script in Github

File: scripts/run-encryption-tests-local.sh [hunk 18]
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env bash
```
Comment (baileympearson) [inline scripts/run-encryption-tests-local.sh]: For both this and the other script - I think adding some comments / documentation explaining what they're for and how they accomplish their tasks would be useful.  MongoDB engineers are familiar with drivers-evergreen-tools, but people outside the company won't be.  Questions like: 

- what is the purpose of drivers-evergreen-tools?
- what it the output / how can you access information about the cluster?
- how can it be configured?

Basically anything that we think somebody who looks at this script might not understand.

File: scripts/run-encryption-tests-local.sh [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+if [ -d "encrypted-cluster" ]; then
+  cd encrypted-cluster
+else
+  source $CWD/scripts/start-encrypted-cluster.sh
+fi
+
+# IMPORTANT: extracts mongodb-uri, and starts the cluster of servers, store the uri for GitHub output
+
+read -r -d '' SOURCE_SCRIPT << EOM
+const fs = require('fs');
```
Comment (baileympearson) [inline scripts/run-encryption-tests-local.sh]: Can we leave a comment explaining what this Node script is doing and why?

File: package.json [hunk 20]
```diff
@@ -104,6 +104,8 @@
     "test-deno": "deno run --allow-env --allow-read --allow-net --allow-run --allow-sys --allow-write ./test/deno.js",
     "test-rs": "START_REPLICA_SET=1 mocha --timeout 30000 --exit ./test/*.test.js",
     "test-tsd": "node ./test/types/check-types-filename && tsd",
+    "test-encryption": "mocha --exit ./test/encryption/*.test.js",
+    "test-encryption-local": "chmod +x scripts/run-encryption-tests-local.sh && scripts/run-encryption-tests-local.sh",
```
Comment (baileympearson) [inline package.json]:
```suggestion
    "test-encryption-local": "bash scripts/run-encryption-tests-local.sh",
```

Usually, we'd just invoke the script with bash instead of chmod'ing it and invoking it directly

**Review by baileympearson — CHANGES_REQUESTED** [review]
@aditi-khare-mongoDB Along with the changes in this PR, can you also put more context about the PR into the PR description?  if/when Val reviews this PR, he likely won't understand why we're making these changes without additional context

File: test/encryption/encryption.test.js (hunk 15 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: that's true! I hadn't considered that

File: test/encryption/encryption.test.js (hunk 13 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: Sure thing!

File: test/encryption/encryption.test.js (hunk 14 above)
Comment (aditi-khare-mongoDB) [reply test/encryption/encryption.test.js]: Sure, changed it to `unencryptedClient`

File: scripts/start-encrypted-cluster.sh (hunk 16 above)
Comment (aditi-khare-mongoDB) [reply scripts/start-encrypted-cluster.sh]: removed it, but why don't we need to input the node driver anymore?

File: scripts/run-encryption-tests-local.sh (hunk 17 above)
Comment (aditi-khare-mongoDB) [reply scripts/run-encryption-tests-local.sh]: the portion about extracting the mongodb-uri + cryptd shared path is. I've updated the comment to reflect this.

File: scripts/run-encryption-tests-local.sh (hunk 19 above)
Comment (aditi-khare-mongoDB) [reply scripts/run-encryption-tests-local.sh]: I've updated the comments for this file!

File: package.json (hunk 20 above)
Comment (aditi-khare-mongoDB) [reply package.json]: Sounds good, made the change!

File: scripts/run-encryption-tests-local.sh (hunk 18 above)
Comment (aditi-khare-mongoDB) [reply scripts/run-encryption-tests-local.sh]: Ok I've added some more clarifying information!

File: mongocryptd.pid
Comment (baileympearson) [reply mongocryptd.pid:1]: Can we add pid files to the git ignore?

File: scripts/start-encrypted-cluster.sh [hunk 21]
```diff
@@ -0,0 +1,36 @@
+# creates a encrypted cluster (sharded on 8.0 server)
+
+export CWD=$(pwd);
+mkdir encrypted-cluster
+cd encrypted-cluster
+
+# note: 
+  # we're using drivers-evergreen-tools which is a repo that handles cluster set-up for us.
```
Comment (baileympearson) [inline scripts/start-encrypted-cluster.sh]:
```suggestion
  # we're using drivers-evergreen-tools which is a repo used by MongoDB drivers to start clusters for testing.  
  # Note: drivers-evergreen-tools is NOT intended to start and manage production servers.
```

we should be explicit that this is only intended to be used for testing

File: scripts/run-encryption-tests-local.sh [hunk 22]
```diff
@@ -0,0 +1,37 @@
...
+
+# set up encrypted mongodb cluster if the encrypted-cluster folder does not exist
+# note: for tooling, cluster set-up and configuration look into the 'scripts/start-encrypted-cluster.sh' script
+if [ -d "encrypted-cluster" ]; then
+  cd encrypted-cluster
+else
+  source $CWD/scripts/start-encrypted-cluster.sh
+fi
+
+# extracts MONGOOSE_TEST_URI and CRYPT_SHARED_LIB_PATH from .yml file into environment variables for this test run
```
Comment (baileympearson) [inline scripts/run-encryption-tests-local.sh]: Not quite - this script doesn't load the files into the environment.  Instead, it reformats the yaml file into bash syntax (`export VAR=<value>`).  the variables are loaded into the environment when we call `source expansions.sh`

File: scripts/run-encryption-tests-local.sh [hunk 23]
```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env bash
+
+# sets up an encrypted mongodb cluster, adds relevant variables to the environment, and runs encryption tests
```
Comment (baileympearson) [inline scripts/run-encryption-tests-local.sh]: (this applies everywhere throughout this PR) there isn't a such thing as an "encrypted cluster".  clusters that use FLE are still just regular mongod/mongoses, but there are some additional setup steps we need to do to use FLE.  We must
 
1. have mongocryptd or the shared library downloaded
2. have an enterprise server >= 4.2

Additionally, I'm pretty sure that as the server changes, the shared library does too - so we should use the same crypt_shared as was included in the server download.

--

drivers-evergreen-tools handles all of this for us - so all we need to do is run the bash script and we get:

- an enterprise server running with the configuration that we chose (enterprise by default, I think you can change that)
- mongocryptd downloaded
- crypt_shared downloaded

We could use other setup tools, but it's just simplest to use drivers-evergreen-tools to set up the cluster because that's what we use.

Could we update the language here to be a bit more precise?  For example, instead of saying "encrypted cluster", perhaps we can say "downloads all tools required to use FLE with the driver / mongoose and and starts a cluster of the provided configuration".

**Review by baileympearson — CHANGES_REQUESTED** [review]
A few small additional comments!

File: scripts/run-encryption-tests-local.sh (hunk 22 above)
Comment (aditi-khare-mongoDB) [reply scripts/run-encryption-tests-local.sh]: To clarify, I was referring to the rest of the code before the next comment, not just the script! So I figured this was okay, what do you think?

File: scripts/start-encrypted-cluster.sh (hunk 21 above)
Comment (aditi-khare-mongoDB) [reply scripts/start-encrypted-cluster.sh]: Since we're not in a mongodb driver right now, but in mongoose, I'd rather not just say drivers-evergreen-tools is just 'used by mongodb drivers'.

File: scripts/run-encryption-tests-local.sh (hunk 23 above)
Comment (aditi-khare-mongoDB) [reply scripts/run-encryption-tests-local.sh]: That's a good point - I've pushed up the other changes you've requested as well. 


Tracking note: Bailey's on leave starting last Friday and I was directed to focus on client.close() work so I didn't push up the small request changes until now.

File: scripts/start-encrypted-cluster.sh (hunk 21 above)
Comment (baileympearson) [reply scripts/start-encrypted-cluster.sh]: I don't feel strongly about whether or not we mention drivers specifically, but we need to emphasize that this is only intended to testing and not an officially supported product for managing clusters.

File: scripts/start-encrypted-cluster.sh (hunk 16 above)
Comment (baileympearson) [reply scripts/start-encrypted-cluster.sh]: It's not used in this script or in the script that I use locally.  Even if it were, this path wouldn't be accurate unless we also had a copy of the driver cloned at `~/dev/node-mongodb-native` wherever this script is running.

File: scripts/run-encryption-tests-local.sh (hunk 22 above)
Comment (baileympearson) [reply scripts/run-encryption-tests-local.sh]: Sure, that's fine too.  I just wanted to clarify what the script does.

File: .github/workflows/encryption-tests.yml (hunk 2 above)
Comment (baileympearson) [reply .github/workflows/encryption-tests.yml]: Do we want to be able to trigger the action manually, or is on "push" / "pr to master" sufficient?

File: scripts/start-encrypted-cluster.sh (hunk 21 above)
Comment (aditi-khare-mongoDB) [reply scripts/start-encrypted-cluster.sh]: might be easier to just leave 'used by drivers' to make that point clear!

File: .github/workflows/encryption-tests.yml (hunk 2 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: I think manually would be nice especially as we continue encryption testing development and if Val or another mongoose contributor adds any encryption functionality.

File: .github/workflows/encryption-tests.yml [hunk 24]
```diff
@@ -0,0 +1,47 @@
...
+      - name: Setup node
+        uses: actions/setup-node@0a44ba7841725637a19e28fa30b79a866c81b0a6 # v4.0.4
+        with:
+          node-version: latest
+      - name: Install Dependencies
+        run: npm install
+      - name: Install mongodb-client-encryption
+        run: npm install mongodb-client-encryption
+      - name: Set up cluster
+        id: setup-cluster
```
Comment (baileympearson) [inline .github/workflows/encryption-tests.yml]: Now that we have tooling to run encryption tests using drivers-evergreen-tools - can we just use the same tooling in CI instead of using the github action?

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: .github/workflows/encryption-tests.yml (hunk 24 above)
Comment (aditi-khare-mongoDB) [reply .github/workflows/encryption-tests.yml]: Good call!

**Review by baileympearson — APPROVED** [review]
